File: nodejs/version2_traduccion.js
// http://localhost:3001/?n=10

const http = require('http');

const SOAP_URL = 'https://www.dataaccess.com/webservicesserver/NumberConversion.wso';

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url, 'http://localhost:3001');
  const n = url.searchParams.get('n');

  if (!n) {
    res.end('Proporciona un número: http://localhost:3001/?n=10');
    return;
  }

  try {
    const soap = `<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
    <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
      <ubiNum>${n}</ubiNum>
    </NumberToWords>
  </soap:Body>
</soap:Envelope>`;

    // Llamada SOAP
    const soapResp = await fetch(SOAP_URL, {
      method: 'POST',
      headers: { 
        'Content-Type': 'text/xml; charset=utf-8',
        'SOAPAction': '""'
      },
      body: soap
    });

    const soapText = await soapResp.text();
    console.log('Respuesta SOAP:', soapText.substring(0, 300));

    // Parsear respuesta con o sin namespace
    let enIngles = '';
    if (soapText.includes('NumberToWordsResult>')) {
      enIngles = soapText
        .split('NumberToWordsResult>')[1]
        .split('<')[0]
        .trim();
    } else {
      res.end('Error: no se encontró resultado en: ' + soapText.substring(0, 200));
      return;
    }

    console.log('En inglés:', enIngles);

    // Traducción
    const tradResp = await fetch(
      `https://api.mymemory.translated.net/get?q=${encodeURIComponent(enIngles)}&langpair=en|es`
    );
    const tradJson = await tradResp.json();
    const traducido = tradJson.responseData.translatedText;

    res.setHeader('Content-Type', 'text/plain; charset=utf-8');
    res.end(traducido);

  } catch (err) {
    console.log('Error completo:', err);
    res.end('Error: ' + err.message);
  }
});

server.listen(3001, () => console.log('Servidor en http://localhost:3001/?n=10'));

File: golang/version2_traduccion.go
// http://localhost:9091/?n=10
// go run version2_traduccion.go

package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		n := r.URL.Query().Get("n")
		if n == "" {
			fmt.Fprint(w, "Proporciona un número: http://localhost:9091/?n=10")
			return
		}

		// Llamada SOAP
		soap := `<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
    <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
      <ubiNum>` + n + `</ubiNum>
    </NumberToWords>
  </soap:Body>
</soap:Envelope>`

		resp, err := http.Post(
			"https://www.dataaccess.com/webservicesserver/NumberConversion.wso",
			"text/xml; charset=utf-8",
			strings.NewReader(soap),
		)
		if err != nil {
			http.Error(w, err.Error(), 500)
			return
		}
		defer resp.Body.Close()

		body, _ := io.ReadAll(resp.Body)
		bodyStr := string(body)
		enIngles := strings.TrimSpace(
			strings.Split(strings.Split(bodyStr, "NumberToWordsResult>")[1], "<")[0],
		)

		// Traducción con MyMemory API
		apiURL := "https://api.mymemory.translated.net/get?q=" +
			url.QueryEscape(enIngles) + "&langpair=en%7Ces"

		tradResp, err := http.Get(apiURL)
		if err != nil {
			http.Error(w, err.Error(), 500)
			return
		}
		defer tradResp.Body.Close()

		tradBody, _ := io.ReadAll(tradResp.Body)

		var resultado map[string]interface{}
		json.Unmarshal(tradBody, &resultado)

		responseData := resultado["responseData"].(map[string]interface{})
		traducido := responseData["translatedText"].(string)

		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		fmt.Fprint(w, traducido)
	})

	fmt.Println("Servidor en http://localhost:9091/?n=10")
	http.ListenAndServe(":9091", nil)
}
